main/pipeline/pipeline.cc

```diff
@@ -899,7 +853,8 @@ ast::ParsedFilesOrCancelled resolve(unique_ptr<core::GlobalState> &gs, vector<as
                     auto reresolved =
                         pipeline::incrementalResolve(*gs, move(toBeReResolved), foundHashesForFiles, opts);
                     ENFORCE(reresolved.size() == 1);
-                    f = checkNoDefinitionsInsideProhibitedLines(*gs, move(reresolved[0]), 0, prohibitedLines);
+                    sorbet::pipeline::definition_checker::checkNoDefinitionsInsideProhibitedLines(
```

Comment:
Despite this being fast enough to replace with trees, I don't think it's going to be fast enough that we want to do it multiple times.

Let's make a set of FileRef and pass that into checkNoDefinitions instead of a single file ref, and then move the call to checkNoDefinitions outside of this loop.

i.e., there should only be one loop over global state, not one loop over global state per file in the test.

Reply:
When you do that, you'll discover that you need a way to know for a given file, what lines are prohibited.

My suggestion: stop using lines for this entirely, and do it all in terms of character offets.

There's a clever realization: for every file we want to check definitions' locs, the length of the file will always be `2 * n + 1` if it started out with length `n`, because we add `n + 1` newlines to the file.

So you can compute the "midpoint" (the first offset at which it's valid to have a location) by doing `(n / 2) + 1`, and then assert that `midpoint <= loc.beginPos() && midpoint <= loc.endPos()`.

That way the only arguments that checkNoDefinitions needs is a GlobalState and an `UnorderedSet<core::FileRef>`.

Reply:
This will also have the effect of making the logic faster, because we won't have to compute the linebreaks of the file at all (which is slow, because it scans the whole file)—we'll do everything in terms of simple offsets.

It will also mean that we need neither of `offset2Pos` nor `position` anymore.